File: Infrastructure/UserContext/Permissions/PermissionService.cs
using Infrastructure.DomainValidation;
using Infrastructure.DomainValidation.Models.ErrorCodes;
using Infrastructure.Permissions.Constants;

namespace Infrastructure.Permissions
{
    public class PermissionService
    {
        private readonly UserContext userContext;
        private readonly DomainValidatorService domainValidatorService;

        public PermissionService(
            UserContext userContext,
            DomainValidatorService domainValidatorService
        )
        {
            this.userContext = userContext;
            this.domainValidatorService = domainValidatorService;
        }

        public void VerifyIsCompanyException()
        {
            if (!VerifyIsCompanyUser())
            {
                domainValidatorService.ThrowErrorMessage(PermissionErrorCode.Permission_NotEnoughPermissions);
            }
        }

        public bool VerifyIsCompanyUser()
        {
            return userContext.OrganizationalUnits
                            .Any(e => e.Alias == OrganizationalUnitConstants.CompanyAlias && e.ExternalId.HasValue);
        }

        public void VerifyUnitPermissionException(string permission, List<(string, int?)> unitExternalIds)
        {
            if (!VerifyUnitPermission(permission, unitExternalIds))
            {
                domainValidatorService.ThrowErrorMessage(PermissionErrorCode.Permission_NotEnoughPermissions);
            }
        }

        public bool VerifyUnitPermission(string permission, List<(string, int?)> unitExternalIds)
        {
            return userContext.OrganizationalUnits
                .Any(e => unitExternalIds.Any(s => s.Item1 == e.Alias 
                    && s.Item2 == e.ExternalId
                    && (string.IsNullOrWhiteSpace(permission) || e.Permissions.Contains(permission))));
        }

        public void VerifyOrganizationalUnitsAliasException(List<string> unitAliases)
        {
            if (!VerifyOrganizationalUnitsAlias(unitAliases))
            {
                domainValidatorService.ThrowErrorMessage(PermissionErrorCode.Permission_NotEnoughPermissions);
            }
        }

        public bool VerifyOrganizationalUnitsAlias(List<string> unitAliases)
        {
            return userContext.OrganizationalUnits.Any(e => unitAliases.Contains(e.Alias));
        }

        public void VerifyPermissionException(string permission)
        {
            if (!VerifyPermission(permission))
            {
                domainValidatorService.ThrowErrorMessage(PermissionErrorCode.Permission_NotEnoughPermissions);
            }
        }

        public bool VerifyPermission(string permission)
        {
            return userContext.OrganizationalUnits.Any(s => s.Permissions.Contains(permission));
        }
    }
}

File: WebApplication/src/auth/services/permission.service.ts
import { Injectable } from "@angular/core";
import { UserContextService } from "./user-context.service";
import { companyAlias } from "../constants/organizational-unit.constants";

@Injectable()
export class PermissionService {

    constructor(private userContextService: UserContextService) {
    }

    verifyIsCompanyUserWithPermission(permission: string) {
        return this.userContextService.userContext.organizationalUnits
            .some(e => e.alias === companyAlias && e.externalId && (!permission || e.permissions.indexOf(permission) > -1));
    }

    verifyUnitPermission(permission: string, unitExternalIds: Array<[string, number]>) {
        return this.userContextService.userContext.organizationalUnits
            .some(e => unitExternalIds
                .some(s => s[0] == e.alias
                    && s[1] == e.externalId
                    && (!permission || e.permissions.indexOf(permission) > -1)));
    }

    verifyOrganizationalUnitsAlias(unitAliases: string[]) {
        return this.userContextService.userContext.organizationalUnits.some(e => unitAliases.indexOf(e.alias) > -1);
    }

    verifyPermission(permission: string) {
        return this.userContextService.userContext.organizationalUnits.some(s => s.permissions.indexOf(permission) > -1);
    }
}